src/main/java/com/redhat/devtools/lsp4ij/features/navigation/LSPGotoDeclarationHandler.java

```diff
@@ -42,25 +52,75 @@ public class LSPGotoDeclarationHandler implements GotoDeclarationHandler {
 
     @Nullable
     @Override
-    public PsiElement[] getGotoDeclarationTargets(@Nullable PsiElement sourceElement, int offset, Editor editor) {
+    public PsiElement[] getGotoDeclarationTargets(@Nullable PsiElement sourceElement,
+                                                  int offset,
+                                                  Editor editor) {
         Project project = editor.getProject();
         if (project == null || project.isDisposed()) {
             return PsiElement.EMPTY_ARRAY;
         }
-        PsiFile psiFile = sourceElement.getContainingFile();
+        PsiFile psiFile = sourceElement != null ? sourceElement.getContainingFile() : null;
         if (psiFile == null) {
             return PsiElement.EMPTY_ARRAY;
         }
         if (!LanguageServersRegistry.getInstance().isFileSupported(psiFile)) {
             return PsiElement.EMPTY_ARRAY;
         }
+
+        // If this was called for a populated semantic tokens view provider, try to get the target directly from it
+        LSPSemanticTokensFileViewProvider semanticTokensFileViewProvider = LSPSemanticTokensFileViewProvider.getInstance(psiFile);
+        if (semanticTokensFileViewProvider != null) {
+            // If this is definitely a reference, resolve it
+            if (semanticTokensFileViewProvider.isReference(offset)) {
+                PsiReference reference = semanticTokensFileViewProvider.findReferenceAt(offset);
+                PsiElement target = reference != null ? reference.resolve() : null;
+                return target != null ? new PsiElement[]{target} : PsiElement.EMPTY_ARRAY;
+            }
+            // If it's definitely a declaration, just return an empty set of targets
+            else if (semanticTokensFileViewProvider.isDeclaration(offset)) {
+                return PsiElement.EMPTY_ARRAY;
+            }
+        }
+
+        // Use LSP to find targets
+        PsiElement[] targets = getGotoDeclarationTargets(editor, sourceElement, offset);
+
+        // If this is a semantic token-backed file and there were targets but this wasn't represented in semantic tokens
+        // as a reference, stub a reference for the word at the current offset
+        if (semanticTokensFileViewProvider != null) {
+            if (!ArrayUtil.isEmpty(targets)) {
```

Comment:
If we got targets for which we didn't have a reference from semantic tokens, stub one for the current word.